delfin/drivers/dell_emc/vplex/vplex_stor.py

```diff
@@ -324,16 +324,221 @@ def get_cluster_used_capacity(self, cluster_name):
+                if director_name:
+                    if "What:" in detail:
+                        match_obj = re.search(r'What:\s+(.+?)$', detail)
+                        if match_obj:
+                            version_name = match_obj.group(1)
+                        continue
+                    if version_name:
+                        match_obj = re.search(r'Version:\s+(.+?)$', detail)
+                        if match_obj:
+                            version_value = match_obj.group(1)
+                            if director_version_map.get(director_name):
+                                director_version_map.get(director_name)[
+                                    version_name] = version_value
+                            else:
+                                version_map = {}
+                                version_map[version_name] = version_value
+                                director_version_map[
+                                    director_name] = version_map
+
+    @staticmethod
+    def analyse_director_status(status):
+        return consts.CONTROLLER_STATUS_MAP. \
+            get(status, constants.ControllerStatus.UNKNOWN)
+
+    def get_director_specified_version(self, version_map, director_name,
+                                       specified_name):
+        version_value = ''
+        if version_map:
+            director_map = version_map.get(director_name)
+            if director_map:
+                version_value = director_map.get(specified_name)
+        return version_value
+
+    def get_value_from_nest_map(self, nest_map, first_key, second_key):
+        final_value = ''
+        if nest_map:
+            second_map = nest_map.get(first_key)
+            if second_map:
+                final_value = second_map.get(second_key)
+        return final_value
+
+    def get_hardware_port_info(self, nest_map, first_key, second_key):
+        speed = ''
+        max_speed = ''
+        protocols = []
+        role = ''
+        port_status = ''
+        operational_status = ''
+
+        if nest_map:
+            second_map = nest_map.get(first_key)
+            if second_map:
+                third_map = second_map.get(second_key)
+                if third_map:
+                    speed = third_map.get('current-speed')
+                    max_speed = third_map.get('max-speed')
+                    protocols = third_map.get('protocols')
+                    role = third_map.get('role')
+                    port_status = third_map.get('port-status')
+                    operational_status = third_map.get('operational-status')
+        return (speed, max_speed, protocols, role, port_status,
+                operational_status)
+
+    @staticmethod
+    def analyse_hardware_port(resp, hardware_port_map):
+        port_list = VplexStorageDriver.get_context_list(resp)
+        if port_list:
+            for port in port_list:
+                port_name = port.get("attributes").get("target-port")
```

Comment:
Suggest not do like that `map.get('xxx').get('yyy')`,  because `map.get('xxx')` may be `None`

Reply:
Modified